<a href="https://colab.research.google.com/github/theshashank1/cse-c5-3year-da/blob/main/tips.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tips Dataset: Statistical Research Notebook

This notebook explores tipping behavior using the `tips` dataset. We will use statistical methods to determine if factors like smoking status, gender, or the day of the week significantly impact bills and tips.

In [8]:
import pandas as pd
from scipy import stats

# Load the dataset
df = pd.read_csv('/content/tips.csv')
display(df.head())

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


In [9]:
df = pd.read_csv('/content/tips.csv')

## 1. Data Exploration and Feature Engineering

We calculate a new feature, `tip_pct` (tip percentage), to normalize the tip amount relative to the size of the total bill.

In [10]:
df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


The `df.head()` command shows the first few rows of the DataFrame, giving us a glimpse of the data. We can see columns like `total_bill`, `tip`, `sex`, `smoker`, `day`, `time`, and `size`.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


The `df.info()` command provides a summary of the DataFrame, including the number of non-null entries for each column and their data types. This helps us identify any missing values or incorrect data types at a glance.

In [12]:
df['tip_pct'] = (df['tip'] / df['total_bill']) * 100
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
 7   tip_pct     244 non-null    float64
dtypes: float64(3), int64(1), object(4)
memory usage: 15.4+ KB


## 2. Feature Engineering: Calculating Tip Percentage

To better understand tipping behavior, we'll create a new feature: `tip_pct`, which represents the tip amount as a percentage of the total bill.

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
 7   tip_pct     244 non-null    float64
dtypes: float64(3), int64(1), object(4)
memory usage: 15.4+ KB


After adding the new `tip_pct` column, we can see it reflected in the DataFrame's information. It's a `float64` type, as expected.

In [14]:
df.head(25)

,total_bill,tip,sex,smoker,day,time,size,tip_pct
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673
1,10.34,1.66,Male,No,Sun,Dinner,3,16.054159
2,21.01,3.50,Male,No,Sun,Dinner,3,16.658734
3,23.68,3.31,Male,No,Sun,Dinner,2,13.978041
4,24.59,3.61,Female,No,Sun,Dinner,4,14.680765
5,25.29,4.71,Male,No,Sun,Dinner,4,18.623962
6,8.77,2.00,Male,No,Sun,Dinner,2,22.805017
7,26.88,3.12,Male,No,Sun,Dinner,4,11.607143
8,15.04,1.96,Male,No,Sun,Dinner,2,13.031915
9,14.78,3.23,Male,No,Sun,Dinner,2,21.853857


Looking at the first 25 rows again, we can now see the calculated `tip_pct` for each entry.

In [15]:
smoker_pct = df[df['smoker'] == 'Yes']['tip_pct']
non_smoker_pct = df[df['smoker'] == 'No']['tip_pct']

t_stat, p_val = stats.ttest_ind(smoker_pct, non_smoker_pct)
print(f'T-statistic: {t_stat:.4f}')
print(f'P-value: {p_val:.4f}')

T-statistic: 0.4797
P-value: 0.6319


## 2. Hypothesis Testing: Smokers vs. Non-Smokers

**Goal:** Determine if smokers tip a different percentage than non-smokers.

*   **H0 (Null Hypothesis):** No difference in mean tip % between groups.
*   **H1 (Alternative Hypothesis):** There is a significant difference.

In [16]:
# T-test

t_stat, p_val = stats.ttest_ind(smoker_pct, non_smoker_pct)
print(f't-stat : {t_stat}, p-val : {p_val}')

t-stat : 0.4796693002669833, p-val : 0.6318957776878545


Now, we perform the independent samples t-test. The `stats.ttest_ind` function from `scipy` calculates the t-statistic and the p-value.

**Interpretation:**
*   **t-statistic:** Measures the difference between the group means in units of standard error.
*   **p-value:** The probability of observing a t-statistic as extreme as, or more extreme than, the one calculated, assuming the null hypothesis is true.

With a p-value of `0.63189`, which is much greater than a common significance level (e.g., 0.05), we **fail to reject the null hypothesis**. This suggests that there is no statistically significant difference in tip percentages between smokers and non-smokers in this dataset.

In [17]:
male_pct = df[df['sex'] == 'Male']['tip_pct']
female_pct = df[df['sex'] == 'Female']['tip_pct']

t_stat, p_val = stats.ttest_ind(male_pct, female_pct, equal_var=False)
print(f'P-value: {p_val:.4f}')

P-value: 0.2542


## 3. Hypothesis Testing: Gender Influence

**Goal:** Does the gender of the payer affect the tip percentage?

*   **H0:** Gender has no effect on tip %.
*   **H1:** Gender significantly affects tip %.

In [18]:
print(len(df[df['sex'] == 'Male']))
print(len(df[df['sex'] == 'Female']))

157
87


The p-value of `0.2542` is greater than 0.05, so we **fail to reject the null hypothesis**. This indicates that there is no statistically significant difference in tip percentages between male and female customers in this dataset.

Let's check the number of male and female customers in our dataset to understand the sample sizes for these groups.

In [19]:
days = ['Thur', 'Fri', 'Sat', 'Sun']
groups = [df[df['day'] == d]['total_bill'] for d in days]

f_stat, p_val = stats.f_oneway(*groups)
print(f'F-statistic: {f_stat:.4f}')
print(f'P-value: {p_val:.4f}')

F-statistic: 2.7675
P-value: 0.0425


## 4. ANOVA: Comparing Bills Across Days

**Goal:** Test if the average bill amount changes depending on the day of the week.

*   **H0:** The mean total bill is the same for all days.
*   **H1:** At least one day has a significantly different mean bill.

In [20]:
for g in groups:
  print(g.mean(), g.std(), len(g))

17.682741935483868 7.886170333087846 62
17.15157894736842 8.302659723217143 19
20.441379310344825 9.480418641086723 87
21.41 8.832121828869889 76


Before running ANOVA, it's good practice to inspect the descriptive statistics (mean, standard deviation, and count) for each group.

In [21]:
f_stat, p_val = stats.f_oneway(*groups)
print(f'f-stat : {f_stat}, p-val : {p_val}')

f-stat : 2.767479443286335, p-val : 0.04245383328952047


Now, we perform the one-way ANOVA test using `stats.f_oneway`.

**Interpretation:**
*   **F-statistic:** The ratio of the variance between the groups to the variance within the groups.
*   **p-value:** The probability of observing an F-statistic as extreme as, or more extreme than, the one calculated, assuming the null hypothesis is true.

With a p-value of `0.0424`, which is less than 0.05, we **reject the null hypothesis**. This suggests that there is a statistically significant difference in the mean `total_bill` among at least some of the days of the week.

In [22]:
contingency = pd.crosstab(df['time'], df['smoker'])
contingency

smoker,No,Yes
time,,
Dinner,106,70
Lunch,45,23


## 5. Chi-Squared Test: categorical Relationships

We use a Chi-Squared test to see if categorical variables like `Time of Day` and `Smoking Status` are related.

In [23]:
contingency = pd.crosstab(df['time'], df['smoker'])
chi2, p_val, dof, expected = stats.chi2_contingency(contingency)

print(f'Chi2 P-value: {p_val:.4f}')
display(pd.DataFrame(expected, columns=contingency.columns, index=contingency.index))

Chi2 P-value: 0.4771


smoker,No,Yes
time,,
Dinner,108.918033,67.081967
Lunch,42.081967,25.918033


**Hypotheses:**
*   **Null Hypothesis (H0):** Smoking status and meal time are independent (no relationship).
*   **Alternative Hypothesis (H1):** Smoking status and meal time are dependent (there is a relationship).

Now, we perform the Chi-squared test for independence using `stats.chi2_contingency`.

In [24]:
pd.DataFrame(expected, columns=contingency.columns, index=contingency.index)

smoker,No,Yes
time,,
Dinner,108.918033,67.081967
Lunch,42.081967,25.918033


With a p-value of `0.4771`, which is greater than 0.05, we **fail to reject the null hypothesis**. This indicates that there is no statistically significant relationship between smoking status and meal time in this dataset.

The expected frequencies show what we would expect if there was no relationship between the two variables.

In [25]:
contingency = pd.crosstab(df['sex'], df['smoker'])
contingency

smoker,No,Yes
sex,,
Female,54,33
Male,97,60


## 7. Chi-squared Test: Gender and Smoking Status

Finally, let's examine if there's a relationship between `sex` and `smoker` status using another Chi-squared test.

First, we create a contingency table for `sex` and `smoker`.

In [26]:
chi2, p_val, dof, expected = stats.chi2_contingency(contingency)
print(f'chi2 : {chi2}, p-val : {p_val}, dof : {dof}')

chi2 : 0.0, p-val : 1.0, dof : 1


**Hypotheses:**
*   **Null Hypothesis (H0):** Gender and smoking status are independent (no relationship).
*   **Alternative Hypothesis (H1):** Gender and smoking status are dependent (there is a relationship).

Now, we perform the Chi-squared test.

In [27]:
pd.DataFrame(expected, columns=contingency.columns, index=contingency.index)

smoker,No,Yes
sex,,
Female,53.840164,33.159836
Male,97.159836,59.840164


With a p-value of `1.0`, which is much greater than 0.05, we **fail to reject the null hypothesis**. This suggests that there is no statistically significant relationship between gender and smoking status in this dataset.

Here are the expected frequencies based on the assumption of independence.

## Summary of Results

1.  **Smokers vs Non-Smokers:** P > 0.05. No significant difference found.
2.  **Gender:** P > 0.05. No significant difference found.
3.  **Day of Week:** P < 0.05. **Significant difference** in how much people spend depending on the day.
4.  **Time vs Smoker:** P > 0.05. No significant relationship.